### DISCUSSION TITLE
help with creating function to match columns based on common string, and sum these columns into a new column in the data frame

### DISCUSSION THREAD

**Discussion by liftr98** [community/community#63789] [Programming Help]
### Body

Hi Everyone,

Im looking for a solution to a problem I have been staring at for too long.

I have a data frame that contains an id variable, and then data for these individuals, taken at many follow ups.

For example. I have physical activity data for different sports. I need to take each sport and then sum all instances, and add it to the end of the data frame.  

Example data:

running1 <- c(2,5,4,7,8,9,4,3,2,6)
running2 <-c(4,5,8,9,3,7,8,3,2,1)
running3<- c(2,5,4,7,8,9,4,3,2,6)
walking1 <- c(2,5,4,7,8,9,4,3,2,6)
walking2 <-c(2,5,4,7,8,9,4,3,2,6)
walking3<-c(2,5,4,7,8,9,4,3,2,6)
id<- c(01,02,03,04,05,06,07,08,09,10)
physical_activity<- data.frame(id, running1, running2, running3, walking1, walking2, walking3)

The way I have gone about it is to get a vector of the column names from the data frame, and then tried to create a function that takes this vector, and for each string in the vector, finds the columns in the data frame that contain this string, and then rowsum. Then selects this summed column and the id variable and join back to the data frame.  

The attempts have looked something like this
 
  ```
 sum_function<- function(colnames_vector,data_frame){
     b<- str_extract(colnames_vector, "[a-z]+")

    z<- data_frame %>% 
      select(n_eid, contains(b))
    z$newvar<- rowSums(z[2:4], na.rm = TRUE) 
    colnames(z$newvar)<-b 
    z<-z %>% select(c(1,5))
    
    data_frame<-data_frame %>% 
      left_join(z, by = "id")}

    sapply(colnames_vector,sum_function(data_frame = physical_activity))

```
I've also tried using for loops, but the function would not maintain the string name this way. Im open to completely changing the method if anyone has an easier way using just the data frame as the input and matching columns within the df. 

I want the output to look like this (just with the data in the running and walking column being the summed values from the corresponding columns)

running1 <- c(2,5,4,7,8,9,4,3,2,6)
running2 <-c(4,5,8,9,3,7,8,3,2,1)
running3<- c(2,5,4,7,8,9,4,3,2,6)
walking1 <- c(2,5,4,7,8,9,4,3,2,6)
walking2 <-c(2,5,4,7,8,9,4,3,2,6)
walking3<-c(2,5,4,7,8,9,4,3,2,6)
id<- c(01,02,03,04,05,06,07,08,09,10)
running<- c(1,1,1,1,1,1,1,1,1,1)
walking<- c(1,1,1,1,1,1,1,1,1,1)
physical_activity<- data.frame(id, running1, running2, running3, walking1, walking2, walking3, running, walking)

If anyone has any ideas it would be much appreciated! 

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by oftharsh** [comment]
It looks like you're trying to aggregate the `running` and `walking` columns based on similar column names in your `physical_activity` data frame. You want to sum up the values from columns with the same name pattern and add those sums as new columns to the data frame. Here's how you can achieve this using the `dplyr` package in R:

```R
library(dplyr)

# Your original data
running1 <- c(2,5,4,7,8,9,4,3,2,6)
running2 <-c(4,5,8,9,3,7,8,3,2,1)
running3 <- c(2,5,4,7,8,9,4,3,2,6)
walking1 <- c(2,5,4,7,8,9,4,3,2,6)
walking2 <- c(2,5,4,7,8,9,4,3,2,6)
walking3 <- c(2,5,4,7,8,9,4,3,2,6)
id <- c(01,02,03,04,05,06,07,08,09,10)
physical_activity <- data.frame(id, running1, running2, running3, walking1, walking2, walking3)

# Sum the columns with the same name pattern and add them to the data frame
physical_activity <- physical_activity %>%
  mutate(running = rowSums(select(., starts_with("running")), na.rm = TRUE),
         walking = rowSums(select(., starts_with("walking")), na.rm = TRUE))

# Print the result
print(physical_activity)
```

This code uses the `mutate` function from the `dplyr` package to calculate the sums of columns with names starting with "running" and "walking" for each row and adds those sums as new columns `running` and `walking` to your `physical_activity` data frame.

Remember to install and load the `dplyr` package using `install.packages("dplyr")` and `library(dplyr)` before running the code.

**Comment by liftr98** [comment]
> 
Ah I didn't know I could use the select function within rowsums! this is great . Thanks a lot for your help.
